lib/auth/init.go

```diff
@@ -281,6 +278,36 @@ func Init(cfg InitConfig, dynamicConfig bool) (*AuthServer, *Identity, error) {
 	return asrv, identity, nil
 }
 
+// setClusterName will set the cluster name resource if it doesn't exist on
+// the backend. if it does exist on the backend and differs from what is in
+// the configuration file we will hardfail. if it matches, we do nothing.
+func setClusterName(cfg InitConfig, asrv *AuthServer) error {
+	cn, err := asrv.GetClusterName()
+	if err != nil && !trace.IsNotFound(err) {
+		return trace.Wrap(err)
+	}
+
+	// if nothing is set on the backend, set and return
+	if trace.IsNotFound(err) {
+		err = asrv.SetClusterName(cfg.ClusterName)
```

Comment:
does it have `upsert` semantics? if yes, there could be a race condition here. Probably should use `CreateClusterName` and simply read again if it was already created and move on

Reply:
This all occurs under a lock: https://github.com/gravitational/teleport/blob/master/lib/auth/init.go#L121-L125

Reply:
I would like to eventually get rid of this backend lock and to do this we should use either compare and swap or optimistic update everywhere and not rely on it in the code, makes sense?

Reply:
Sure, I'll make the change.